nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+  if (ith > num_node - 1)
+    throw std::invalid_argument("Exceed total number of layer");
+
+  adj[ith].push_back(node);
+}
+
+void NetworkGraph::addLayerNode(std::shared_ptr<Layer> layer) {
+  std::list<LayerNode> l;
+  LayerNode *node = new LayerNode();
+
+  node->layer = layer;
+  node->index = num_node;
+
+  l.assign(1, *node);
+  adj.push_back(l);
+  num_node++;
+}
+
+void NetworkGraph::topologicalSortUtil(unsigned int ith, bool visited[],
+                                       std::stack<LayerNode> &Stack) {
+  visited[ith] = true;
+
+  std::list<LayerNode>::iterator i;
+  for (i = adj[ith].begin(); i != adj[ith].end(); ++i) {
+    if (!visited[(*i).index])
+      topologicalSortUtil((*i).index, visited, Stack);
+  }
+
+  Stack.push(getLayerNode(ith));
+}
+
+LayerNode &NetworkGraph::getLayerNode(unsigned int ith) {
+
+  std::list<LayerNode>::iterator iter;
+  for (unsigned int i = 0; i < num_node; ++i) {
+    iter = adj[i].begin();
+    if ((*iter).index == ith) {
+      return *iter;
+    }
+  }
+
+  throw std::invalid_argument("Cannot find Layer");
+}
+
+void NetworkGraph::topologicalSort() {
+  std::stack<LayerNode> Stack;
+
+  bool *visited = new bool[num_node];
+  for (unsigned int i = 0; i < num_node; ++i)
+    visited[i] = false;
+
+  for (unsigned int i = 0; i < num_node; ++i) {
+    if (visited[i] == false) {
+      topologicalSortUtil(i, visited, Stack);
+    }
+  }
+
+  delete[] visited;
+
+  while (Stack.empty() == false) {
+    Sorted.push_back(Stack.top());
+    Stack.pop();
+  }
+
+  for (unsigned int i = 0; i < Sorted.size(); ++i) {
+    std::cout << Sorted[i].layer->getName() << " ";
+  }
+  std::cout << std::endl;
+}
+
+void NetworkGraph::ensureName(std::shared_ptr<Layer> layer,
+                              const std::string &prefix = "",
+                              bool force_rename = false) {
+  std::string orig_name = layer->getName();
+  bool orig_name_empty = orig_name.empty();
+  if (!orig_name_empty && !force_rename &&
+      layer_names.end() == layer_names.find(orig_name))
+    return;
+
+  /** If just prefix with layer name makes it unique - directly set the name */
+  if (!orig_name_empty) {
+    std::string direct_name = prefix + orig_name;
+    if (layer_names.find(direct_name) == layer_names.end()) {
+      layer->setName(direct_name);
+      return;
+    }
+  }
+
+  std::set<std::string>::iterator iter;
+  std::string name;
+  if (orig_name_empty)
+    orig_name = layer->getType();
+  std::string direct_name = prefix + orig_name;
+
+  do {
+    name = direct_name + std::to_string(def_name_count++);
+    iter = layer_names.find(name);
+  } while (iter != layer_names.end());
+
+  layer->setName(name);
+}
+
+int NetworkGraph::realizeMultiInputType(Layer &current) {
+  int status = ML_ERROR_NONE;
+  if (current.num_inputs == 1)
+    return ML_ERROR_NONE;
+
+  if (current.num_inputs > 1) {
+    std::shared_ptr<Layer> layer = nntrainer::createLayer("concat");
```

Comment:
how does the graph choose between concat and addition layer in case of multi input?
if its NYI, please add a TODO.

Reply:
I think it should be `nntrainer::createLayer(layer->getType())`